ergo/distributions/point_density.py

```diff
@@ -0,0 +1,260 @@
+        if not normalized:
+            xs = scale.normalize_points(xs)
+            densities = scale.normalize_densities(xs, densities)
+
+        if allow_non_standard_pairs:
+            # interpolate ps at target_xs
+            if not (
+                len(xs) == len(constants.target_xs)
+                and np.isclose(xs, constants.target_xs, rtol=1e-04).all()
+            ):
+                f = interp1d(xs, densities)
+                densities = f(constants.target_xs)
+
+        # Make sure AUC is 1
+        auc = np.sum(densities) / densities.size
+        densities /= auc
+
+        return cls(constants.target_xs, densities, scale=scale, normalized=True)
+
+    @classmethod
+    def from_conditions(
+        cls,
+        *args,
+        fixed_params=None,
+        scale=None,
+        num_points=constants.point_density_default_num_points,
+        **kwargs,
+    ):
+
+        if fixed_params is None:
+            # TODO: Seems weird to denormalize when will be normalized in normalize_fixed_params
+            fixed_params = {"xs": scale.denormalize_points(constants.target_xs)}
+        else:
+            raise Exception("Point Density does not accept custom xs specifications")
+
+        return super(PointDensity, cls).from_conditions(
+            *args, fixed_params=fixed_params, scale=scale, **kwargs
+        )
+
+    @classmethod
+    def from_params(cls, fixed_params, opt_params, scale=Scale(0, 1), traceable=True):
+        xs = fixed_params["xs"]
+        ps = nn.softmax(opt_params) * opt_params.size
+        densities = ps
+        return cls(
+            xs=xs, densities=densities, scale=scale, normalized=True, traceable=True
+        )
+
+    # Optimize Create Helpers
+
+    @staticmethod
+    def initialize_optimizable_params(fixed_params):
+        num_points = fixed_params["xs"].size
+        return onp.full(num_points, 1.0)
+
+    @classmethod
+    def normalize_fixed_params(self, fixed_params, scale):
+        return {"xs": scale.normalize_points(fixed_params["xs"])}
+
+    # Export
+
+    def to_arrays(self, metaculus_denorm=False):
+        xs = self.scale.denormalize_points(self.normed_xs)
+
+        if metaculus_denorm:
+            # Make sure points cover whole scale, return normed densities
+            densities = self.normed_densities
+            if xs[0] != self.scale.low:
+                density = (densities[0] - densities[1]) / 2 + densities[0]
+                clamped_density = onp.maximum(density, 0)
+
+                xs = onp.append(onp.array([self.scale.low]), xs)
+                densities = onp.append(np.array([clamped_density]), densities)
+            if xs[-1] != self.scale.high:
+                density = (densities[-1] - densities[-2]) / 2 + densities[-1]
+                clamped_density = onp.maximum(density, 0)
+
+                xs = onp.append(xs, onp.array([self.scale.high]))
+                densities = onp.append(densities, np.array([clamped_density]))
```

Comment:
Could factor this out as a function and describe and test the logic separate from PointDensity.

Reply:
Factored out into `add_endpoints` and added a test in `test_point_density`